src/Lib.hs

```diff
@@ -0,0 +1,145 @@
+module Lib
+    ( CDNum(..)
+    , CDOps(..)
+    , simplify
+    , constructFromList
+    ) where
+
+
+{-|
+https://pl.wikipedia.org/wiki/Aksjomaty_i_konstrukcje_liczb#Konstrukcja_Cayleya-Dicksona
+
+* Para liczb rzeczywistych jest liczbą zespoloną
+* Para liczb zespolonych jest Kwaternionem
+* Para kwaternionów jest oktonionem
+* Można tak w nieskończoność
+
+Ta klasa opisuje działania które można wykonać na takich typach
+Potrzeba ją definiować dla typu podstawowego (jak liczby rzeczywiste w powyższym przykładzie)
+-}
+class (Fractional a) => CDOps a where
+  neutralAdd :: a
+  neutralMul :: a
+  conjugate :: a -> a
+  normSqr :: a -> a
+  normSqr a = a * (conjugate a)
+
+instance (CDOps a) => Num (CDNum a) where
+  (+) = cdAdd
+  negate = cdNegative
+  (*) = cdMultiply
+  fromInteger a = Base (fromInteger a)
+  signum = cdSignum
+  abs = cdNormSqr --todo pierwiastek
+  
+instance (CDOps a) => Fractional (CDNum a) where
+  recip = cdRecip
+  fromRational a = Base (fromRational a)
+  
+instance (CDOps a) => CDOps (CDNum a) where
+  neutralAdd = cdNeutralAdd
+  neutralMul = cdNeutralMultiply
+  conjugate = cdConjugate
+  normSqr = cdNormSqr
+
+-- |Patrz: 'CDOps'.
+data CDNum a = Base a | Pair (CDNum a)  (CDNum a) deriving (Eq)
+instance (Show a) => Show (CDNum a) where
+  show (Base a) = show a
+  show (Pair a b) = "["++(show a) ++" "++ (show b) ++ "]"
+
+
+--Przykładowym CDOpsem będzie Double
+instance CDOps Double where
+  neutralAdd = 0
+  neutralMul = 1
+  conjugate a = a
+
+
+-- Poniżej są już tylko szczegóły implementacji
+
+{-|
+@
+[1] -> [1]
+[1 2] -> [1 2]
+[1 2 3 4] -> [[1 2] [3 4]]
+[1 2 3 4 5 6 7 8] -> [[1 2 3 4] [5 6 7 8]] -> [[[1 2] [3 4]] [[5 6] [7 8]]]
+@
+-}
+constructFromList :: (CDOps a) => [a] -> CDNum a
+constructFromList lst = constructFromCDNumList  (map Base lst)
+--todo: w razie gdyby lista nie była długości 2^n trzeba będzie dopełnić zerami
```

Comment:
coś w stylu:
let len = length lst
     n = (2^) . ceiling . logBase 2 . fromIntegral $ len
     k = n - len
in constructFromCDNumList . map Base . concat $ [lst, take k . repeat $ 0]